README.rst
Trac Redis Pub
==============
`Trac <https://trac.edgewall.org/>`_ plugin providing `Redis pub/sub
channels <https://redis.io/topics/pubsub>`_ for Trac events, specifically
ticket and wiki page creation/updates.

This can be used to implement services external to the Trac web server
itself that react to events on a Trac project.  For example, one can provide
a stream of ticket events like:

.. code:: python

    >>> import redis
    >>> client = redis.Redis()
    >>> pubsub = client.pubsub()
    >>> pubsub.psubscribe('trac.*')
    >>> for event in pubsub.listen():
    ...     print(event)
    ...
    {'pattern': None, 'type': 'psubscribe', 'channel': 'trac.*', 'data': 1L}
    {'pattern': 'trac.*', 'type': 'pmessage', 'channel': 'trac.test.wiki.created', 'data': '{"comment": "Change comment", "name": "NewWikiPage", "author": "anonymous", "text": "Page contents.", "readonly": 0, "version": 1, "time": "2018-03-02T14:10:22.844985+00:00"}'}
    {'pattern': 'trac.*', 'type': 'pmessage', 'channel': 'trac.test.wiki.changed.NewWikiPage', 'data': '{"comment": "Editing page.", "name": "NewWikiPage", "author": "anonymous", "text": "Page contents.\\r\\nAdditional contents.", "old_text": "Page contents.", "readonly": 0, "version": 2, "time": "2018-03-02T14:10:36.192988+00:00"}'}
    {'pattern': 'trac.*', 'type': 'pmessage', 'channel': 'trac.test.ticket.created', 'data': '{"status": "new", "changetime": "2018-03-02T14:15:01.401989+00:00", "reporter": "anonymous", "cc": "", "milestone": "", "component": "component1", "keywords": "", "owner": "somebody", "id": 17, "description": "Problem description.", "author": "", "summary": "Test ticket", "priority": "major", "version": "", "time": "2018-03-02T14:15:01.401989+00:00", "type": "defect"}'}
    {'pattern': 'trac.*', 'type': 'pmessage', 'channel': 'trac.test.ticket.changed.17', 'data': '{"comment": "Updated milestone.", "new_values": {"status": "new", "changetime": "2018-03-02T14:15:01.401989+00:00", "reporter": "anonymous", "cc": "", "milestone": "milestone1", "component": "component1", "keywords": "", "owner": "somebody", "id": 17, "description": "Problem description.", "author": "", "summary": "Test ticket", "priority": "major", "version": "", "time": "2018-03-02T14:15:01.401989+00:00", "type": "defect"}, "id": 17, "old_values": {"milestone": ""}, "author": "anonymous"}'}
    

.. note::

    Version 0.1 supports basic ticket and wiki-related events.  However,
    there are many other types of events in Trac that could be monitored
    this way, for which support may be added in future versions in the
    remote chance there is any demand.


Installation and Configuration
------------------------------

The plugin can be installed in the standard way using ``pip``::

    $ pip install tracext-redispub

To enable the plugin in Trac, add the following to the ``[components]``
section of your ``trac.ini``:

.. code:: ini

    [components]
    tracext.redispub.redis.redisclient = enabled
    tracext.redispub.ticket.redisticketstream = enabled
    tracext.redispub.wiki.rediswikistream = enabled

The ``tracext.redispub.redis.redisclient`` component must be enabled for the
other components to work.  The other components may be enabled or disabled
as desired (e.g. to disable all wiki channels use ``tracext.redispub.wiki.*
= disabled``).

The plugin is further configured through two additional configuration
sections.  The first section is ``[redis_client]`` which is used for
configuring how to connect to the Redis server.  Currently this only takes a
small number of options--later it will be expanded to the fuller range of
options for configuring your Redis connection:

.. code:: ini

    [redis_client]
    # hostname of the Redis server
    host = localhost
    # Redis server port
    port = 6379
    # Redis DB number (although this can be configured, it is irrelevant for
    # pub/sub purposes)
    db = 0
    # Redis server password
    password =
    # Path to UNIX socket to connect over instead of through TCP
    unix_socket_path =

One additional section configures the behavior of the plugin itself (and
currently has only one option):

.. code:: ini

    [redispub]
    # This string is prefixed to the names of all channels published to by
    # the plugin
    channel_prefix = trac


Available Channels
------------------

The following channels can be subscribed to:

Ticket channels
^^^^^^^^^^^^^^^

* Ticket creation events are published to the
  ``<prefix>.<env>.ticket.created`` channel, where ``<prefix>`` is the
  configurable channel name prefix (e.g. 'trac'), and ``<env>`` is the
  environment name.

  Messages on this channel consist of the field values of the created ticket
  (along with the ticket ID) as a JSON-encoded dictionary:

  .. code:: json
  
      {
          "id": 1,
          "summary": "...",
          "description": "...",
          ...
      }

* Ticket change events are published to the
  ``<prefix>.<env>.ticket.changed.<id>`` channel, where ``<prefix>`` and
  ``<env>`` are as before, and ``<id>`` is the ticket ID.  This allows
  subscribing just to the changes on a specific ticket, if desired.  The
  message is a JSON-encoded dictionary with the following format:

  .. code:: json
  
      {
          "id": 1,
          "new_values": { ... },
          "old_values": { ... },
          "author": "somebody",
          "comment": "A comment..."
      }

  Where ``"id"`` is the ticket ID. ``"new_values"`` maps field names to
  their new values (including fields that did not change), and
  ``"old_values"`` maps field names to the previous values of fields that
  changed.  ``"author"`` is the author of the change, and ``"comment"`` is
  the comment associated with the change (which may be blank).

* Ticket deletion events are published to the
  ``<prefix>.<env>.ticket.deleted`` channel.  These events have the same
  format as ticket creation events, and include the values of all the fields
  on the just-deleted ticket.

Wiki channels
-------------

* Wiki page creation events are published to the
  ``<prefix>.<env>.wiki.created`` channel, where ``<prefix>`` is the
  configurable channel name prefix (e.g. 'trac'), and ``<env>`` is the
  environment name.

  Messages on this channel are JSON-encoded dictionaries representing the
  wiki page (``"version"`` is always ``1`` for new pages):

  .. code:: json

      {
          "name": "PageName",
          "version": 1,
          "time": "2018-03-02T12:31:28.184283",
          "author": "somebody",
          "text": "...full page text...",
          "comment": "edit comment, if any",
          "readonly": 0
      }

* Wiki change events are published to the
  ``<prefix>.<env>.wiki.changed.<name>`` channel, where ``<prefix>`` and
  ``<env>`` are as before, and ``<name>`` is the wiki page name.  This
  allows subscribing just to the changes on a specific wiki page, if
  desired.  The message is a JSON-encoded dictionary with the same format as
  the wiki created event, along with an additional ``"old_text"`` property
  containing the previous text of the wiki page before the change:
  
  .. code:: json

      {
          "name": "PageName",
          "version": 2,
          ...
          "text": "the new text of the page",
          "old_text": "the old text of the page",
          ...
      }

* Wiki deletion events are published to the ``<prefix>.<env>.wiki.deleted``
  channel.  These events have the same format as wiki creation events, and
  include the properties of the just-deleted wiki page..



CHANGES.rst
0.2 (unreleased)
================

* Changed the ``ticket.changed`` channel so that the ``"new_values"``
  keyword contains all of the changed ticket's fields, including those that
  did not change.  This allows more flexibility in formatting the ticket
  changed messages.


0.1 (03/02/2018)
================

* Initial release.


tracext/redispub/redis.py
"""Core Redis client component."""

from __future__ import absolute_import


from trac.config import Option, IntOption, PathOption
from trac.core import Component, TracError

import redis


class RedisClient(Component):
    """
    A mini-component for managing a `redis.Redis` client instance (and its
    underlying connection pool) for use by other components that use Redis, as
    well as options passed to the client.
    """

    _config_section = 'redis_client'

    _redis_client_options = {
        'host': 'localhost',
        'port': 6379,
        'db': 0,
        'password': None,
        'unix_socket_path': (None, 'path')
    }


    _type_to_option_getter = {
        type(None): 'get',
        str: 'get',
        int: 'getint',
        'path': 'getpath'
    }

    def __init__(self):
        super(RedisClient, self).__init__()
        self.redis = redis.Redis(**self._client_options())

    def _client_options(self):
        """
        Read supported arguments to `redis.Redis` from the Trac config.
        """

        options = {}
        for key, value in self._redis_client_options.items():
            if isinstance(value, tuple):
                if len(value) == 1:
                    default = value[0]
                    type_ = None
                elif len(value) == 2:
                    default, type_ = value
            else:
                default = value
                type_ = None

            if type_ is None:
                # Infer type from the default type
                type_ = type(default)

            getter = getattr(self.env.config,
                             self._type_to_option_getter[type_])
            options[key] = getter(self._config_section, key, default)

        return options


class RedisComponent(Component):
    """
    A simple base class for `Component`s that provides a ``.redis`` property
    returning the `Redis` client instance of the `RedisClient` component.
    """

    channel_prefix = Option('redispub', 'channel_prefix', 'trac',
                            doc="Prefix to use for channels published to by "
                                "components of this plug-in.  All other "
                                "channels are dotted with this name; e.g. "
                                "trac.<envname>.ticket.created")

    _realm = None
    """
    The realm (e.g. 'ticket', 'wiki') handled by subclasses of this component.
    """

    def __init__(self):
        if not self.env.is_enabled(RedisClient):
            raise TracError(
                "The {0}.{1} component must be enabled in order to use the "
                "{0}.{2} component.".format(__name__, RedisClient.__name__,
                                            self.__class__.__name__))

        super(RedisComponent, self).__init__()

    @property
    def redis(self):
        """Returns the `Redis` client instance."""

        return self.env[RedisClient].redis

    def _channel_name(self, method):
        channel = method
        if self._realm is not None:
            channel = self._realm + '.' + channel
        channel = self.env.name + '.' + channel
        if self.channel_prefix:
            channel = self.channel_prefix + '.' + channel
        return channel


tracext/redispub/ticket.py
"""Ticket event handling."""

from __future__ import absolute_import


from trac.core import implements
from trac.ticket.api import ITicketChangeListener

from .redis import RedisComponent
from .util import dumps


class RedisTicketStream(RedisComponent):
    """
    Listens for ticket creation/change/deletion events and and publishes
    them to the relevant channels:

        * Ticket creation events are published to the
          ``<prefix>.<env>.ticket.created`` channel, where ``<prefix>`` is the
          configurable channel name prefix (e.g. 'trac'), and ``<env>`` is the
          environment name.

          Messages on this channel consist of the field values of the created
          ticket (along with the ticket ID) as a JSON-encoded dictionary::

              {
                  "id": 1,
                  "summary": "...",
                  "description": "...",
                  ...
              }

        * Ticket change events are published to the
          ``<prefix>.<env>.ticket.changed.<id>`` channel, where ``<prefix>``
          and ``<env>`` are as before, and ``<id>`` is the ticket ID.  This
          allows subscribing just to the changes on a specific ticket, if
          desired.  The message is a JSON-encoded dictionary with the following
          format::

              {
                  "id": 1,
                  "new_values": { ... },
                  "old_values": { ... },
                  "author": "somebody",
                  "comment": "A comment..."
              }

          Where ``"id"`` is the ticket ID. ``"new_values"`` maps field names to
          their new values (including fields that did not change), and
          ``"old_values"`` maps field names to the previous values of fields
          that changed.  ``"author"`` is the author of the change, and
          ``"comment"`` is the comment associated with the change (which may be
          blank).

        * Ticket deletion events are published to the
          ``<prefix>.<env>.ticket.deleted`` channel.  These events have the
          same format as ticket creation events, and include the values of all
          the fields on the just-deleted ticket.
    """

    implements(ITicketChangeListener)

    _realm = 'ticket'

    # ITicketChangeListener methods
    def ticket_created(self, ticket):
        data = dict(ticket.values)
        data['id'] = ticket.id
        self.redis.publish(self._channel_name('created'), dumps(data))

    def ticket_changed(self, ticket, comment, author, old_values):
        data = {
            'new_values': ticket.values,
            'old_values': old_values,
            'author': author,
            'comment': comment,
            'id': ticket.id,
        }
        self.redis.publish(self._channel_name('changed', ticket.id),
                           dumps(data))

    def ticket_deleted(self, ticket):
        data = dict(ticket.values)
        data['id'] = ticket.id
        self.redis.publish(self._channel('deleted'), dumps(data))

    def _channel_name(self, method, ticket_id=None):
        channel = super(RedisTicketStream, self)._channel_name(method)
        if ticket_id is not None:
            channel += '.' + str(ticket_id)
        return channel


tracext/redispub/wiki.py
"""Wiki event handling."""

from __future__ import absolute_import


from trac.core import implements
from trac.wiki.api import IWikiChangeListener

from .redis import RedisComponent
from .util import dumps


class RedisWikiStream(RedisComponent):
    """
    Listens for wiki creation/change/deletion events and and publishes
    them to the relevant channels:

        * Wiki page creation events are published to the
          ``<prefix>.<env>.wiki.created`` channel, where ``<prefix>`` is the
          configurable channel name prefix (e.g. 'trac'), and ``<env>`` is the
          environment name.

          Messages on this channel are JSON-encoded dictionaries representing
          the wiki page (``"version"`` is always ``1`` for new pages)::

              {
                  "name": "PageName",
                  "version": 1,
                  "time": "2018-03-02T12:31:28.184283",
                  "author": "somebody",
                  "text": "...full page text...",
                  "comment": "edit comment, if any",
                  "readonly": 0
              }

        * Wiki change events are published to the
          ``<prefix>.<env>.wiki.changed.<name>`` channel, where ``<prefix>``
          and ``<env>`` are as before, and ``<name>`` is the wiki page name.  This
          allows subscribing just to the changes on a specific wiki page, if
          desired.  The message is a JSON-encoded dictionary with the same
          format as the wiki created event, along with an additional
          ``"old_text"`` property containing the previous text of the wiki page
          before the change::

              {
                  "name": "PageName",
                  "version": 2,
                  ...
                  "text": "the new text of the page",
                  "old_text": "the old text of the page",
                  ...
              }

        * Wiki deletion events are published to the
          ``<prefix>.<env>.wiki.deleted`` channel.  These events have the same
          format as wiki creation events, and include the properties of the
          just-deleted wiki page..
    """

    implements(IWikiChangeListener)

    _realm = 'wiki'

    # ITicketChangeListener methods
    def wiki_page_added(self, page):
        self.redis.publish(self._channel_name('created'),
                           self._page_to_json(page))

    def wiki_page_changed(self, page, version, t, comment, author, ipnr=None):
        # note: the version, timestamp (t), comment, and author arguments
        # are all passed to this method for backwards-compatibility, but the
        # same data is already available in the page object
        self.redis.publish(self._channel_name('changed', page.name),
                           self._page_to_json(page, old_text=True))

    def wiki_page_deleted(self, ticket):
        self.redis.publish(self._channel_name('deleted'),
                           self._page_to_json(page))

    def _channel_name(self, method, page_name=None):
        channel = super(RedisWikiStream, self)._channel_name(method)
        if page_name is not None:
            channel += '.' + page_name
        return channel

    def _page_to_json(self, page, old_text=False):
        """
        Convert a `trac.wiki.model.WikiPage` object to a JSONified dictionary.
        """
        attrs = ['name', 'version', 'time', 'author', 'text', 'comment',
                 'readonly']

        if old_text:
            attrs.append('old_text')

        return dumps(dict((attr, getattr(page, attr)) for attr in attrs))
